### PR TITLE
Factor out time management code

### PR DESCRIPTION
**Pull request by mooskagh**
Not sure it made code better, not worse, but the intention was the following:

Factor out time management code from search/engine to a separate directory in order to:
1. Be able to write better time manager (e.g. carry nps from one search to another, change time budget as the search goes, etc).
2. Move complexity out of Search and Engine classes.
3. Have reusable part of the code which can be used for search v2.

It should be mostly no-op, but there are some differences, e.g.
- remaining time and nodes are stored per thread, not globally, that means that threads may sometimes disagree on that.
- `lc0 benchmark` doesn't support smart pruning now.

I'll probably add tests as the change is large and messay.

### CODE REVIEW

**Comment by mooskagh** [comment]
Instead of passing limits to a search and search itself watching for it, this change makes it work the opposite:

1. "Stopper" is passed to the search.
2. Search calls stopper after every iteration, passing needed stats.
3. Stopper returns whether search should be stopped (and also passes "hints": estimated nodes remaining, and estimated time remaining)

**Comment by oscardssmith** [comment]
This change seems great. Getting unnecessary complication out of search is highly appreciated.

**Comment by borg323** [comment]
`lc0 benchmark` not supporting smart pruning is probably an improvement.

**Comment by gonzalezjo** [comment]
> Not sure it made code better, not worse

Seems like it's better overall.

File: src/mcts/timemgr/stoppers.cc [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+}
+
+int64_t TimeLimitStopper::GetTimeLimitMs() const { return time_limit_ms_; }
+
+///////////////////////////
+// DepthStopper
+///////////////////////////
+bool DepthStopper::ShouldStop(const IterationStats& stats, TimeManagerHints*) {
+  if (stats.average_depth >= depth_) {
+    LOGFILE << "Stopped search: Reached depth.";
```
Comment (borg323) [inline src/mcts/timemgr/stoppers.cc]: return true;

Comment (mooskagh) [reply src/mcts/timemgr/stoppers.cc]: Done.

File: src/mcts/stoppers/stoppers.cc [hunk 2]
```diff
@@ -0,0 +1,235 @@
+/*
+  This file is part of Leela Chess Zero.
+  Copyright (C) 2018-2019 The LCZero Authors
```
Comment (borg323) [inline src/mcts/stoppers/stoppers.cc]: Is this intentional? All other files in this directory mention only 2019.

**Pull request merged**